from snakemake.utils import min_version

##### set minimum snakemake version #####
min_version("7.31.1")

# define variables
transcript_gtf = config["transcripts"]
samples = config["samples"]
ref_genome = config["ref_genome"]
out_dir = config["out_dir"]
sp_fasta = config["swissprot_fasta"]
diamond_dir = config["diamond_db_dir"]
diamond_db_prefix = config["diamond_db_prefix"]

# load rules
include: "rules/predict_ORFs.smk"
include: "rules/visualize_ORFs.smk"

rule all:
    input:
        expand(os.path.join(out_dir, "proteome", "{sample}", "proteins.fa"), sample=samples),
        expand(os.path.join(out_dir, "visualization", "{sample}", "transcripts.bed"), sample=samples),
        expand(os.path.join(out_dir,"visualization","{sample}","transcripts.fasta.transdecoder.genome.bed"),
            sample=samples)


